### PR TITLE
uneven splitting of codepoints

### PR DESCRIPTION
**Pull request by markroxor**
refer #36

### CODE REVIEW

**Comment by markroxor** [comment]
Not the final patch, I'm just experimenting. I'll ping you once it is done.

**Comment by markroxor** [comment]
@jarun just gave it a quick fix, can you confirm on your machine?

File: nnn.c [hunk 1]
```diff
@@ -1274,7 +1274,18 @@ printent_long(struct entry *ent, int sel)
 		else if (ent->mode & 0100)
 			snprintf(g_buf, ncols, "%s%-16.16s %8.8s* %s*", CURSYM(sel), buf, coolsize(ent->size), unescape(ent->name));
 		else
-			snprintf(g_buf, ncols, "%s%-16.16s %8.8s  %s", CURSYM(sel), buf, coolsize(ent->size), unescape(ent->name));
+        {
+            // experimenting in this section
+            const char *c = unescape(ent->name);
+
+            int count = 30;
```
Comment (jarun) [inline nnn.c]: What's this count? Why 30?

File: nnn.c [hunk 2]
```diff
@@ -1274,7 +1274,18 @@ printent_long(struct entry *ent, int sel)
 		else if (ent->mode & 0100)
 			snprintf(g_buf, ncols, "%s%-16.16s %8.8s* %s*", CURSYM(sel), buf, coolsize(ent->size), unescape(ent->name));
 		else
-			snprintf(g_buf, ncols, "%s%-16.16s %8.8s  %s", CURSYM(sel), buf, coolsize(ent->size), unescape(ent->name));
+        {
+            // experimenting in this section
```
Comment (jarun) [inline nnn.c]: Is this the final code or a test code?

File: nnn.c [hunk 3]
```diff
@@ -1274,7 +1274,18 @@ printent_long(struct entry *ent, int sel)
...
+            const char *c = unescape(ent->name);
+
+            int count = 30;
+            int goback=0;
```
Comment (jarun) [inline nnn.c]: Please maintain spacing (e.g. `goback = 0`) everywhere. I tend to follow the linux kernel coding style in my projects (other that the 80 char limit per line):

https://01.org/linuxgraphics/gfx-docs/drm/process/coding-style.html

Comment (jarun) [reply nnn.c]: This applies to the rest of the changes too.

File: nnn.c [hunk 4]
```diff
@@ -1274,7 +1274,18 @@ printent_long(struct entry *ent, int sel)
...
+            if(ncols+goback>count)
+                while((ncols+goback)>count && mblen(c+(ncols+goback)-count, MB_CUR_MAX)==-1)
+                    goback--;
+
+            snprintf(g_buf, ncols+goback, "%s%-16.16s %8.8s %s", CURSYM(sel), buf, coolsize(ent->size), unescape(ent->name));
```
Comment (jarun) [inline nnn.c]: This case takes care of only regular files. Our problem is with names notwithstanding the type of file.

**Comment by jarun** [comment]
@markroxor in addition to my review comments, a general comment (as you have removed the `work in progress` label):

We tend to push only **_production_ready_** code. Your code, unfortunately, is far from that. Many of the utilities are available in multiple Linux distros and OSes. So when we make a release lots of people use them. Which is why we do _NOT_ release hacks or non-optimal patches without a valid reason.

File: nnn.c (hunk 1 above)
Comment (jarun) [reply nnn.c]: OK, I got it from your PR comment. Please leave these code related comments in the diff as general comments. Makes it easier to find.

We can't limit to 30. In this case we can fall back to the non-detailed view mode.

File: nnn.c [hunk 5]
```diff
@@ -1274,7 +1274,18 @@ printent_long(struct entry *ent, int sel)
...
 			snprintf(g_buf, ncols, "%s%-16.16s %8.8s* %s*", CURSYM(sel), buf, coolsize(ent->size), unescape(ent->name));
 		else
-			snprintf(g_buf, ncols, "%s%-16.16s %8.8s  %s", CURSYM(sel), buf, coolsize(ent->size), unescape(ent->name));
+        {
+            // experimenting in this section
+            const char *c = unescape(ent->name);
+
+            int count = 30;
+            int goback=0;
+            if(ncols+goback>count)
```
Comment (jarun) [inline nnn.c]: Similar changes should go into the function `printent()` (non-detail mode) too.

**Comment by jarun** [comment]
@markroxor in addition to my review comments, a general comment (as you have removed the `work in progress` label):

We tend to push only **_production_ready_** code. Your code, unfortunately, is far from that. Many of the utilities are available in multiple Linux distros and OSes. So when we make a release lots of people use them. Which is why we do _NOT_ release hacks or non-optimal patches without a valid reason.

File: nnn.c (hunk 1 above)
Comment (jarun) [reply nnn.c]: I think the best option would be to check the number of cols in the beginning. If it is < 30 call `printent()` with the same args.

Comment (markroxor) [reply nnn.c]: Instead I am using what `snprintf` is returning. Refer code. What do you think?

File: nnn.c (hunk 2 above)
Comment (markroxor) [reply nnn.c]: final code, waiting for your suggestions.

File: nnn.c (hunk 4 above)
Comment (markroxor) [reply nnn.c]: you mean I should extend this to other "else" clause as well?

File: nnn.c [hunk 6]
```diff
@@ -1186,6 +1186,21 @@ unescape(const char *str)
 	return buffer;
 }
 
+int
+getgoback(struct entry *ent,int sel, int ncols)
+{
+    const char *c = unescape(ent->name);
+
+    // counts the number of columns before it enounters the filename
+	int count = snprintf(g_buf, ncols, "%s ", CURSYM(sel));
```
Comment (jarun) [inline nnn.c]: This is fixed to 30. Using 30 was OK. What I intended to say is why not fallback to non-detail mode when we have less than 30 columns?

File: nnn.c [hunk 7]
```diff
@@ -1197,7 +1212,7 @@ printent(struct entry *ent, int sel)
 		ncols = COLS;
 
 	if (S_ISDIR(ent->mode))
-		snprintf(g_buf, ncols, "%s%s/", CURSYM(sel), unescape(ent->name));
+        snprintf(g_buf, ncols, "%s%s/", CURSYM(sel), unescape(ent->name));
```
Comment (jarun) [inline nnn.c]: Indentation issue. Please use tabs.

File: nnn.c (hunk 6 above)
Comment (markroxor) [reply nnn.c]: > This is fixed to 30. Using 30 was OK.

Isn't this better?
> What I intended to say is why not fallback to non-detail mode when we have less than 30 columns?

On it.

File: nnn.c [hunk 8]
```diff
@@ -1186,6 +1186,21 @@ unescape(const char *str)
...
+int
+getgoback(struct entry *ent,int sel, int ncols)
+{
+    const char *c = unescape(ent->name);
+
+    // counts the number of columns before it enounters the filename
+	int count = snprintf(g_buf, ncols, "%s ", CURSYM(sel));
+
+	int goback = 0;
+    while((ncols + goback) > count && mblen(c + (ncols + goback) - count, MB_CUR_MAX) == -1)
```
Comment (jarun) [inline nnn.c]: Please explain what you are trying to do here. I am unable to understand.

Comment (jarun) [reply nnn.c]: Please start with what `goback` is. :)

Comment (markroxor) [reply nnn.c]: Kind of hard to explain but -
`mblen` returns -1 for multi-byte characters which have not ended. Suppose the character is "ं", it is of "length" `3` so mblen will return a non ''-1'' value only when this char is completely printed. So I decremented the variable `goback` until mblen doesn't return a non "-1" value, this is the point from where breaking the code-point is safe.

Did you try executing this locally? Is this what you ask?

File: nnn.c [hunk 9]
```diff
@@ -1186,6 +1186,21 @@ unescape(const char *str)
...
+        goback--;
```
Comment (jarun) [inline nnn.c]: it's always better to use pre-incr and pre-decr. Ref: https://embeddedgurus.com/stack-overflow/2008/08/efficient-c-tips-3-avoiding-post-increment-decrement/

File: nnn.c (hunk 8 above)
Comment (markroxor) [reply nnn.c]: `goback` is the difference between the current column length and the length where breaking the code-point is safe.

Comment (jarun) [reply nnn.c]: It's makes sense to me now. `goback` can only be negative. That's what I missed earlier.

**Comment by jarun** [comment]
I think the changes look good now. I will run a some tests and pull it in.

**Comment by jarun** [comment]
I did some testing. This isn't working. It seems to work because in the line:

    while((ncols + goback) > count && mblen(c + (ncols + goback) - count, MB_CUR_MAX) == -1)

`c + (ncols + goback) - count` makes an illegal memory access if `ncols + goback - count` is > length of the string. So it _cuts_ the string beyond its end.

Thanks for your efforts. I will have to handle this myself because it needs deeper understanding of how the printing works today and curses printing.

**Comment by jarun** [comment]
Please take a look at https://github.com/jarun/nnn/commit/1e5a0b8c5bc78462956d01c647293a7da594c36f.

My understanding is - these kind of issues involve wide char conversion one way or another. It would have been great to be able to avoid it but we are doing it already at `unescape()` so I mixed both and could throw in some optimization. The conversion is always safe as it guarantees split at a codepoint.

**Comment by markroxor** [comment]
I know this is resolved but -
> `c + (ncols + goback) - count` makes an illegal memory access if `ncols + goback - count` is > length of the string. So it _cuts_ the string beyond its end.

This can be easily fixed by limiting it to the total length of printable string? Isn't it?

**Comment by jarun** [comment]
I tried that already. Didn't work as expected. Try it out.

**Comment by jarun** [comment]
I mean test by resizing the terminal live.

**Comment by jarun** [comment]
Also, it's more accurate but slower to calculate the fixed length for each entry. So I did an approximation there including the newline char at the end.

**Comment by markroxor** [comment]
Okay. Anyways the issue is resolved so no point in further tweaking. Deleting the branch. Thanks.

**Pull request closed without merging**